File: js/201_BitwiseANDofNumbersRange.js
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var rangeBitwiseAnd = function(m, n) {
    if(m === 0) return 0;
    var mark = 1;
    while( m !== n){
        m >>= 1;
        n >>= 1;
        mark <<= 1;
    }
    return mark * m;
};

File: cpp/201_BitwiseANDofNumbersRange.cpp
class Solution {
public:
    int rangeBitwiseAnd(int m, int n) {
        if(m == 0) return 0;
        int mark = 1;
        
        while(m != n){
            m >>= 1;
            n >>= 1;
            mark <<= 1;
        }
        return mark * m;
    }
};

class Solution {
public:
    int rangeBitwiseAnd(int m, int n) {
        return m != n ? rangeBitwiseAnd(m >> 1, n >> 1) * 2 : n;
    }
};